src/mongocxx/test/spec/unified_tests/runner.cpp

```diff
@@ -110,6 +110,41 @@ bsoncxx::document::value make_kms_doc() {
     return kms_doc;
 }
 
+bsoncxx::document::value parse_kms_doc(bsoncxx::document::view_or_value test_kms_doc) {
+    auto kms_values = get_kms_values();
+    auto doc = bsoncxx::builder::basic::document{};
+    auto test_kms_doc_view = test_kms_doc.view();
+    for (const auto& it : test_kms_doc_view) {
+        auto provider = it.key();
+        if (!kms_values[provider]) {
+            FAIL("FAIL: got unexpected KMS provider: " << provider);
+        }
+        auto variables_doc = bsoncxx::builder::basic::document{};
+        auto variables = test_kms_doc_view[provider].get_document().view();
+        for (const auto& i : variables) {
+            auto variable = i.key();
+            auto actual_value = kms_values[provider][variable];
+            if (!kms_values[provider][variable]) {
+                FAIL("FAIL: expecting to find variable: '"
+                     << variable << "' in KMS doc for provider: '" << provider << "'");
+            }
+            switch (actual_value.type()) {
```

Comment:
Suggested change to handle the case when the [input is not a placeholder](https://github.com/mongodb/specifications/blob/baf3724155500acbb557e158b8a5d4334e7f3512/source/unified-test-format/unified-test-format.rst#entity).

> Otherwise, drivers MUST configure the KMS provider with the explicit value of KMS credentials field given in the test file (e.g. kmsProviders: { aws: { accessKeyId: abc, secretAccessKey: def } }). This is to permit testing conditions where invalid KMS credentials are provided.

```suggestion
            bool is_placeholder = false;
            if (i.type() == bsoncxx::type::k_document &&
                i.get_document().value == make_document(kvp("$$placeholder", 1))) {
                is_placeholder = true;
            }
            if (!is_placeholder) {
                // Append value as-is.
                variables_doc.append(kvp(variable, i.get_value()));
                continue;
            }
            // A placeholder was specified. Append the credential from the environment.
            switch (actual_value.type()) {
```

I do not think this behavior is tested. But it can be verified by replacing a `{ "$$placeholder": 1}` with `"foobar"` in `rewrapManyDataKey.json` and expecting an error.